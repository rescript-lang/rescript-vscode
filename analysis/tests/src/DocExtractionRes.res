/***Module level documentation goes here. */

/** This type represents stuff. */
type t = {
  /** The name of the stuff.*/
  name: string,
  /** Whether stuff is online.*/
  online: bool,
}

/** Create stuff.

```rescript example
let stuff = make("My name")
```
*/
let make = name => {
  name,
  online: true,
}

/** Stuff goes offline.*/
let asOffline = (t: t) => {...t, online: false}

/** exotic identifier */
let \"SomeConstant\" = 12

module SomeInnerModule = {
  /*** Another module level docstring here.*/
  type status =
    | /** If this is started or not */ Started(t) | /** Stopped? */ Stopped | /** Now idle.*/ Idle

  /** These are all the valid inputs.*/
  type validInputs = [#something | #"needs-escaping" | #withPayload(int) | #status(status)]

  type callback = (t, ~status: status) => unit
}

module AnotherModule = {
  /*** Mighty fine module here too!*/

  /** This links another module. Neat. */
  module LinkedModule = SomeInnerModule

  /**
  Testing what this looks like.*/
  type callback = SomeInnerModule.status => unit

  let isGoodStatus = (status: SomeInnerModule.status) => status == Stopped

  /** Trying how it looks with an inline record in a variant. */
  type someVariantWithInlineRecords =
    | /** This has inline records...*/
    SomeStuff({
        offline: bool,
        /** Is the user online? */ online?: bool,
      })

  open ReactDOM

  /**Callback to get the DOM root...*/
  type domRoot = unit => Client.Root.t
}

module ModuleWithThingsThatShouldNotBeExported: {
  /*** BROKEN: This docstring isn't picked up Doesn't seem to be parsed at all, no attributes found.*/

  /** The type t is stuff. */
  type t

  /** The maker of stuff!*/
  let make: unit => t
} = {
  /*** Mighty fine module here too!*/
  type t = string
  type x = int
  type f = bool

  let m1 = (x: x) => {
    x + 1
  }

  let m2 = (f: f) =>
    if f {
      true
    } else {
      false
    }

  let make = () => {
    if m2(true) && m1(1) > 2 {
      "1"
    } else {
      "2"
    }
  }
}

// ^dex
