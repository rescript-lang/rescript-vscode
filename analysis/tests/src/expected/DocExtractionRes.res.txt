Documentation extraction src/DocExtractionRes.res
extracting docs for src/DocExtractionRes.res

{
  "name": "DocExtractionRes",
  "docstrings": ["Module level documentation goes here."],
  "items": [
  {
    "id": "DocExtractionRes.t",
    "kind": "type",
    "name": "t",
    "signature": "type t = {name: string, online: bool}",
    "docstrings": ["This type represents stuff."],
    "detail": 
    {
      "kind": "record",
      "items": [{
        "name": "name",
        "optional": false,
        "docstrings": ["The name of the stuff."],
        "signature": "string"
      }, {
        "name": "online",
        "optional": false,
        "docstrings": ["Whether stuff is online."],
        "signature": "bool"
      }]
    }
  }, 
  {
    "id": "DocExtractionRes.make",
    "kind": "value",
    "name": "make",
    "signature": "let make: string => t",
    "docstrings": ["Create stuff.\n\n```rescript example\nlet stuff = make(\"My name\")\n```"]
  }, 
  {
    "id": "DocExtractionRes.asOffline",
    "kind": "value",
    "name": "asOffline",
    "signature": "let asOffline: t => t",
    "docstrings": ["Stuff goes offline."]
  }, 
  {
    "id": "DocExtractionRes.SomeConstant\\",
    "kind": "value",
    "name": "SomeConstant\\",
    "signature": "let SomeConstant\\: int",
    "docstrings": ["exotic identifier"]
  }, 
  {
    "id": "DocExtractionRes.SomeInnerModule",
    "name": "SomeInnerModule",
    "kind": "module",
    "items": [
    {
      "id": "DocExtractionRes.SomeInnerModule.status",
      "kind": "type",
      "name": "status",
      "signature": "type status = Started(t) | Stopped | Idle",
      "docstrings": [],
      "detail": 
      {
        "kind": "variant",
        "items": [
        {
          "name": "Started",
          "docstrings": ["If this is started or not"],
          "signature": "Started(t)"
        }, 
        {
          "name": "Stopped",
          "docstrings": ["Stopped?"],
          "signature": "Stopped"
        }, 
        {
          "name": "Idle",
          "docstrings": ["Now idle."],
          "signature": "Idle"
        }]
      }
    }, 
    {
      "id": "DocExtractionRes.SomeInnerModule.validInputs",
      "kind": "type",
      "name": "validInputs",
      "signature": "type validInputs = [\n  | #\"needs-escaping\"\n  | #something\n  | #status(status)\n  | #withPayload(int)\n]",
      "docstrings": ["These are all the valid inputs."]
    }, 
    {
      "id": "DocExtractionRes.SomeInnerModule.callback",
      "kind": "type",
      "name": "callback",
      "signature": "type callback = (t, ~status: status) => unit",
      "docstrings": []
    }]
  }, 
  {
    "id": "DocExtractionRes.AnotherModule",
    "name": "AnotherModule",
    "kind": "module",
    "items": [
    {
      "id": "DocExtractionRes.LinkedModule",
      "kind": "moduleAlias",
      "name": "LinkedModule",
      "docstrings": ["This links another module. Neat."],
      "items": []
    }, 
    {
      "id": "DocExtractionRes.AnotherModule.callback",
      "kind": "type",
      "name": "callback",
      "signature": "type callback = SomeInnerModule.status => unit",
      "docstrings": ["Testing what this looks like."]
    }, 
    {
      "id": "DocExtractionRes.AnotherModule.isGoodStatus",
      "kind": "value",
      "name": "isGoodStatus",
      "signature": "let isGoodStatus: SomeInnerModule.status => bool",
      "docstrings": []
    }, 
    {
      "id": "DocExtractionRes.AnotherModule.someVariantWithInlineRecords",
      "kind": "type",
      "name": "someVariantWithInlineRecords",
      "signature": "type someVariantWithInlineRecords =\n  | SomeStuff({offline: bool})",
      "docstrings": ["Trying how it looks with an inline record in a variant."],
      "detail": 
      {
        "kind": "variant",
        "items": [
        {
          "name": "SomeStuff",
          "docstrings": ["This has inline records..."],
          "signature": "SomeStuff"
        }]
      }
    }, 
    {
      "id": "DocExtractionRes.AnotherModule.domRoot",
      "kind": "type",
      "name": "domRoot",
      "signature": "type domRoot = unit => ReactDOM.Client.Root.t",
      "docstrings": ["Callback to get the DOM root..."]
    }]
  }, 
  {
    "id": "DocExtractionRes.ModuleWithThingsThatShouldNotBeExported",
    "name": "ModuleWithThingsThatShouldNotBeExported",
    "kind": "module",
    "items": [
    {
      "id": "DocExtractionRes.ModuleWithThingsThatShouldNotBeExported.t",
      "kind": "type",
      "name": "t",
      "signature": "type t",
      "docstrings": ["The type t is stuff."]
    }, 
    {
      "id": "DocExtractionRes.ModuleWithThingsThatShouldNotBeExported.make",
      "kind": "value",
      "name": "make",
      "signature": "let make: unit => t",
      "docstrings": ["The maker of stuff!"]
    }]
  }]
}

